replay/models/kl_ucb.py

```diff
@@ -0,0 +1,154 @@
+        \\operatorname{KL}(p, q)
+        = p \\log\\frac{p}{q} + (1-p)\\log\\frac{1-p}{1-q}
+
+    is the KL-divergence of Bernoulli distribution with parameter :math:`p`
+    from Bernoulli distribution with parameter :math:`q`.
+
+    Being a bit trickier though, the bound shares with UCB the same
+    exploration-exploitation tradeoff dilemma. You may increase the `c`
+    coefficient in order to shift the tradeoff towards exploration or decrease
+    it to set the model to be more sceptical of items with small volume of
+    collected statistics. The authors of the `article
+    <https://arxiv.org/pdf/1102.2490.pdf>`_ though claim `c = 0` to be of the
+    best choice in practice.
+
+
+    As any other RePlay model, KL-UCB takes a log to fit on as a ``DataFrame``
+    with columns ``[user_idx, item_idx, timestamp, relevance]``. Following the
+    procedure above, KL-UCB would see each row as a record of an interaction
+    with ``item_idx`` with positive (relevance = 1) or negative (relevance = 0)
+    outcome. ``user_idx`` and ``timestamp`` are ignored.
+
+    If ``relevance`` column is not of 0/1 initially, then you have to decide
+    what kind of relevance has to be considered as positive and convert
+    ``relevance`` to binary format during preprocessing.
+
+    To provide a prediction, KL-UCB would sample a set of recommended items for
+    each user with probabilites proportional to obtained relevances.
+
+    >>> import pandas as pd
+    >>> data_frame = pd.DataFrame({"user_idx": [1, 2, 3, 3], "item_idx": [1, 2, 1, 2], "relevance": [1, 0, 0, 0]})
+    >>> from replay.utils import convert2spark
+    >>> data_frame = convert2spark(data_frame)
+    >>> model = KL_UCB()
+    >>> model.fit(data_frame)
+    >>> model.predict(data_frame,k=2,users=[1,2,3,4], items=[1,2,3]
+    ... ).toPandas().sort_values(["user_idx","relevance","item_idx"],
+    ... ascending=[True,False,True]).reset_index(drop=True)
+       user_idx  item_idx  relevance
+    0         1         3   1.000000
+    1         1	        2   0.750000
+    2         2	        3   1.000000
+    3         2	        1   0.933013
+    4         3	        3   1.000000
+    5         4	        3   1.000000
+    6         4	        1   0.933013
+
+    """
+
+    def __init__(
+        self,
+        exploration_coef: float = 0.0,
+        sample: bool = False,
+        seed: Optional[int] = None,
+    ):
+        """
+        :param exploration_coef: exploration coefficient
+        :param sample: flag to choose recommendation strategy.
+            If True, items are sampled with a probability proportional
+            to the calculated predicted relevance.
+            Could be changed after model training by setting the `sample`
+            attribute.
+        :param seed: random seed. Provides reproducibility if fixed
+        """
+
+        super().__init__(exploration_coef, sample, seed)
+
+    def _calc_item_popularity(self):
+
+        rhs = math.log(self.full_count) \
```

Comment:
Could you make this variable name longer, e.g. right_bound?

Reply:
ok, no problem